More Detailed Compaction Errors

kevinrr888 opened the pull request:
Closes #4075
Changes:
- CompactionFinalizer
	- Added compactionId to log of compaction failure (tableId is already included from 'ecfs')
- CompactionCoordinator
	- Added extent/tableId to log of compaction failure (compactionId already included)
- CompactableImpl
	- Added extent/tableId to two logs of compaction failure (compactionId already included)
- Compactor
	- Added extent/tableId and compactionId to three of the logs
- ExternalDoNothingCompactor
	- Added extent/tableId and compactionId to log of compaction failure
- DeadCompactionDetector
	- Added extent/tableId to log (compactionId already included)
- FileCompactor
	- Added extent/tableId, input files, output file, iterators, and thread start time to log when error occurs

Potentially still left TODO:
- Any more logs that need to be changed?
- Any of the logs I changed that shouldn't have been?
- Other requested changes?

EdColeman commented on server/base/src/main/java/org/apache/accumulo/server/compaction/FileCompactor.java:
```diff
@@ -279,7 +287,10 @@ public CompactionStats call() throws IOException, CompactionCanceledException {
+      log.error("Compaction error. Compaction info: "
+          + "extent: {}, input files: {}, output file: {}, iterators: {}, start date: {}. Error message: {}",
+          getExtent(), getFileNamesToCompact(), getOutputFileName(), getIterators(),
+          threadStartDate, e.getMessage(), e);
```
Is `e.getMessage()` along with `e` redundant.  If the exception is logged with the stack trace it will include the message as well.

kevinrr888 replied:
Yes, I agree. That is how it was originally logged, so I left those in. Wasn't sure if there was some weird reason for having both. I'll update in my next commit.

keith-turner replied:
It would be redundant, could drop that.

keith-turner commented:
> Any more logs that need to be changed?

I do not know if there are others that need to be changed.  The changes that have been made so far are a really nice improvement so it would be good to get them merged.  Can make a PR if more log stmts are found.

kevinrr888 commented on server/base/src/main/java/org/apache/accumulo/server/compaction/FileCompactor.java:
```diff
@@ -189,6 +191,10 @@ protected String getOutputFile() {
     return outputFile.toString();
   }
 
+  protected String getOutputFileName() {
```
Yeah, I'll just inline both those methods since they're only used once.

kevinrr888 commented:
> > Any more logs that need to be changed?
> 
> I do not know if there are others that need to be changed. The changes that have been made so far are a really nice improvement so it would be good to get them merged. Can make a PR if more log stmts are found.

Thank you. Yeah, I believe I have covered all the logs (as far as I've been able to tell), just wanted to double check if anyone else had any further ideas.

ddanielr commented:
Overall the changes look good. 

The build failures seem to be caused by an issue with the mock `extent` object in CompactorTest.
https://github.com/apache/accumulo/pull/4148/files#diff-34232683deeb79b4d648e68a75995a1f462256013c2d8b4c2a1f0e7a92e666ffR563-R564 

The CompactionITs build fine. 

I checked out your branch and replaced the mock object with an actual tKeyExtent and the code passed, but threw me a warning. 
```
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.powermock.reflect.internal.WhiteboxImpl
```

kevinrr888 commented:
Changed the failing test to use a real TKeyExtent object (not mock), fixed the redundant logging, moved my methods that were just used once to inline.

ddanielr commented on server/compaction-coordinator/src/main/java/org/apache/accumulo/coordinator/CompactionFinalizer.java:
```diff
@@ -128,7 +128,8 @@ private void notifyTserver(Location loc, ExternalCompactionFinalState ecfs) {
             ecfs.getFileSize(), ecfs.getEntries());
       } else if (ecfs.getFinalState() == FinalState.FAILED) {
-        LOG.debug("Notifying tserver {} that compaction {} has failed.", loc, ecfs);
+        LOG.debug("Notifying tserver {} that compaction {} with {} has failed.", loc, ecfs,
+            ecfs.getExternalCompactionId());
```
move ID to right after `compaction` in log message

```suggestion
        LOG.debug("Notifying tserver {} that compaction {} with {} has failed.", loc, cfs.getExternalCompactionId(), 
         ecfs);
```

ddanielr reviewed (approved):
Added a minor suggestion regarding a log message's ID position. Overall these changes are a great improvement
